File: services/web-api/src/main/java/com/github/jeffreyjpz/timestamped_for_f1_web_api/web/v1/circuits/dtos/Circuit.java
package com.github.jeffreyjpz.timestamped_for_f1_web_api.web.v1.circuits.dtos;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonProperty;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor
public class Circuit {

    @JsonAlias({"circuitKey", "circuit_key"})
    @JsonProperty("circuit_key")
    private final String circuitKey;

    @JsonAlias({"circuitName", "circuit_name"})
    @JsonProperty("circuit_name")
    private final String circuitName;

    @JsonAlias({"coordinates"})
    @JsonProperty("coordinates")
    private final List<CircuitLocationCoordinates> coordinates;

    @JsonAlias({"countryCode", "country_code"})
    @JsonProperty("country_code")
    private final String countryCode;

    @JsonAlias({"countryKey", "country_key"})
    @JsonProperty("country_key")
    private final Integer countryKey;

    @JsonAlias({"countryName", "country_name"})
    @JsonProperty("country_name")
    private final String countryName;

    @JsonAlias({"location"})
    @JsonProperty("location")
    private final String location;

    @JsonAlias({"marshalSectors", "marshal_sectors"})
    @JsonProperty("marshal_sectors")
    private final List<CircuitLocation> marshalSectors;

    @JsonAlias({"miniSectors", "mini_sectors"})
    @JsonProperty("mini_sectors")
    private final List<CircuitLocation> miniSectors;

    @JsonAlias({"rotation"})
    @JsonProperty("rotation")
    private final Integer rotation;

    @JsonAlias({"turns"})
    @JsonProperty("turns")
    private final List<CircuitLocation> turns;

    @JsonAlias({"year"})
    @JsonProperty("year")
    private final Integer year;

}

File: dashboard/src/types/circuit.ts
import z from "zod";

export const circuitLocationType = z
    .enum([
        "marshal-sector",
        "mini-sector",
        "turn"
    ]);

const circuitLocation = z
    .object({
        angle: z.number(),
        coordinates: z
            .object({
                x: z.number(),
                y: z.number()
            }),
        length: z.number().nonnegative(),
        number: z.number().int().positive(),
        type: circuitLocationType
    });

export const circuit = z
    .object({
        circuit_key: z.number().int().positive(),
        circuit_name: z.string(),
        coordinates: z
            .array(
                z.object({
                    x: z.number(),
                    y: z.number()
                })
            ),
        country_code: z.string().length(3),
        country_key: z.number().int().positive(),
        country_name: z.string(),
        location: z.string(),
        marshal_sectors: z.array(circuitLocation),
        mini_sectors: z.array(circuitLocation),
        rotation: z.number(),
        turns: z.array(circuitLocation),
        year: z.number().int().nonnegative()
    });

export const circuitRead = circuit
    .pick({
        circuit_key: true,
        year: true
    });

export const marshalSector = z
    .object({
        ...circuitLocation.shape,
        type: z.literal("marshal-sector")
    });

export const marshalSectorRead = marshalSector
    .pick({
        number: true
    })
    .extend(
        circuit.pick({
            circuit_key: true,
            year: true
        })
        .shape
    );

export const miniSector = z
    .object({
        ...circuitLocation.shape,
        type: z.literal("mini-sector")
    });

export const miniSectorRead = miniSector
    .pick({
        number: true
    })
    .extend(
        circuit.pick({
            circuit_key: true,
            year: true
        })
        .shape
    );

export const turn = z
    .object({
        ...circuitLocation.shape,
        type: z.literal("turn")
    });

export const turnRead = turn
    .pick({
        number: true
    })
    .extend(
        circuit.pick({
            circuit_key: true,
            year: true
        })
        .shape
    );

export type CircuitLocationType = z.infer<typeof circuitLocationType>;

export type Circuit = z.infer<typeof circuit>;
export type CircuitRead = z.infer<typeof circuitRead>;

export type MarshalSector = z.infer<typeof marshalSector>;
export type MarshalSectorRead = z.infer<typeof marshalSectorRead>;

export type MiniSector = z.infer<typeof miniSector>;
export type MiniSectorRead = z.infer<typeof miniSectorRead>;

export type Turn = z.infer<typeof turn>;
export type TurnRead = z.infer<typeof turnRead>;